### PR TITLE
Use 2D weights in XCM v3

### PR DESCRIPTION
**Pull request by KiChjang**

### CODE REVIEW

File: xcm/src/v2/mod.rs [hunk 1]
```diff
@@ -885,48 +897,81 @@ impl<RuntimeCall> Instruction<RuntimeCall> {
 
 // TODO: Automate Generation
 impl<RuntimeCall, W: XcmWeightInfo<RuntimeCall>> GetWeight<W> for Instruction<RuntimeCall> {
-	fn weight(&self) -> Weight {
+	fn weight(&self) -> sp_weights::Weight {
```
Comment (gavofyork) [inline xcm/src/v2/mod.rs:900]: Why is this being changed? Other than conversion from XCM v3, XCM v2 module should remain the same, no?

File: xcm/src/v3/traits.rs [hunk 2]
```diff
@@ -205,8 +207,8 @@ impl From<SendError> for Error {
 
 pub type Result = result::Result<(), Error>;
 
-/// Local weight type; execution time in picoseconds.
-pub type Weight = u64;
+/// Default value for the proof size weight component. Set at 64 KB.
+pub const DEFAULT_PROOF_SIZE: u64 = 64 * 1024;
```
Comment (gavofyork) [inline xcm/src/v3/traits.rs:211]: This is only used in the `TryFrom<OldInstruction>`, right? In which case, better put it up there next to it as a private constant.

File: xcm/xcm-builder/src/barriers.rs [hunk 3]
```diff
@@ -91,8 +91,10 @@ impl<T: Contains<MultiLocation>> ShouldExecute for AllowTopLevelPaidExecutionFro
 			i = iter.next().ok_or(())?;
 		}
 		match i {
-			BuyExecution { weight_limit: Limited(ref mut weight), .. } if *weight >= max_weight => {
-				*weight = max_weight;
+			BuyExecution { weight_limit: Limited(ref mut weight), .. }
+				if weight.any_gte(max_weight) =>
```
[the remark this answers is not in the document]
Comment (gavofyork) [reply xcm/xcm-builder/src/barriers.rs:95]: @KiChjang please make a note that a second reviewer should carefully check every one of these comparison functions.

File: xcm/xcm-builder/src/barriers.rs [hunk 4]
```diff
@@ -240,7 +242,8 @@ impl<T: Contains<MultiLocation>> ShouldExecute for AllowExplicitUnpaidExecutionF
 		);
 		ensure!(T::contains(origin), ());
 		match instructions.first() {
-			Some(UnpaidExecution { weight_limit: Limited(m), .. }) if *m >= max_weight => Ok(()),
+			Some(UnpaidExecution { weight_limit: Limited(m), .. }) if m.any_gte(max_weight) =>
```
Comment (gavofyork) [inline xcm/xcm-builder/src/barriers.rs:245]:
```suggestion
			Some(UnpaidExecution { weight_limit: Limited(m), .. }) if m.all_gte(max_weight) =>
```

File: xcm/xcm-builder/src/tests/assets.rs [hunk 5]
```diff
@@ -33,8 +33,13 @@ fn exchange_asset_should_work() {
 		},
 	]);
 	let hash = fake_message_hash(&message);
-	let r = XcmExecutor::<TestConfig>::execute_xcm(Parent, message, hash, 50);
-	assert_eq!(r, Outcome::Complete(40));
+	let r = XcmExecutor::<TestConfig>::execute_xcm(
+		Parent,
+		message,
+		hash,
+		Weight::from_ref_time(50).set_proof_size(DEFAULT_PROOF_SIZE),
```
Comment (gavofyork) [inline xcm/xcm-builder/src/tests/assets.rs:40]: I don't think we should be using `DEFAULT_PROOF_SIZE` here in tests - it is strictly for conversion from XCMv2 weight to XCMv3 weight. If you want some vaguely sensible proof-size constant to use in tests, just make a local declaration.

**Comment by xlc** [comment]
Can we have a new instruction `buy_execution(asset: MultiAsset)` that allow user to provide some tokens, and say use as much as required from those tokens to pay for execution fees, without specify exact ref time & pov requirement? So that frontend doesn't need to figure out the ref time & pov value for an extrinsic and the chain could charge additional fees. e.g. storage deposit fee for smart contract calls.

File: xcm/xcm-builder/src/tests/barriers.rs [hunk 6]
```diff
@@ -47,7 +47,10 @@ fn computed_origin_should_work() {
 		DescendOrigin(Parachain(100).into()),
 		DescendOrigin(PalletInstance(69).into()),
 		WithdrawAsset((Parent, 100).into()),
-		BuyExecution { fees: (Parent, 100).into(), weight_limit: Limited(100) },
+		BuyExecution {
+			fees: (Parent, 100).into(),
+			weight_limit: Limited(Weight::from_ref_time(100)),
```
Comment (gavofyork) [inline xcm/xcm-builder/src/tests/barriers.rs:52]: I would add a new helper function `fn from_dummy(all_fields: u64)` which sets both fields to be equal to `all_fields`, and use that everywhere in these tests.

File: xcm/xcm-builder/src/tests/barriers.rs [hunk 7]
```diff
@@ -167,46 +186,46 @@ fn allow_paid_should_work() {
...
+		BuyExecution { fees, weight_limit: Limited(Weight::from_ref_time(30)) },
 		DepositAsset { assets: AllCounted(1).into(), beneficiary: Here.into() },
 	]);
 
 	let r = AllowTopLevelPaidExecutionFrom::<IsInVec<AllowPaidFrom>>::should_execute(
 		&Parachain(1).into(),
 		paying_message.inner_mut(),
-		30,
-		&mut 0,
+		Weight::from_ref_time(30).set_proof_size(30),
```
Comment (gavofyork) [inline xcm/xcm-builder/src/tests/barriers.rs:219]:
```suggestion
		Weight::from_parts(30, 30),
```

Comment (gavofyork) [reply xcm/xcm-builder/src/tests/barriers.rs:219]: This kind of thing.

File: xcm/xcm-builder/src/tests/basic.rs [hunk 8]
```diff
@@ -82,18 +88,18 @@ fn code_registers_should_work() {
 	]);
 	// Weight limit of 70 is needed.
 	let limit = <TestConfig as Config>::Weigher::weight(&mut message).unwrap();
-	assert_eq!(limit, 70);
+	assert_eq!(limit, Weight::from_ref_time(70).set_proof_size(DEFAULT_PROOF_SIZE));
```
Comment (gavofyork) [inline xcm/xcm-builder/src/tests/basic.rs:91]: Weight::from_parts(70, 70)

Comment (gavofyork) [inline xcm/xcm-builder/src/tests/basic.rs:91]:
```suggestion
	assert_eq!(limit, Weight::from_parts(70, 70));
```

File: xcm/xcm-builder/src/tests/bridging/mod.rs [hunk 9]
```diff
@@ -133,8 +133,12 @@ impl<Local: Get<Junctions>, Remote: Get<Junctions>, RemoteExporter: ExportXcm> S
 		set_exporter_override(price::<RemoteExporter>, deliver::<RemoteExporter>);
 		// The we execute it:
 		let hash = fake_message_hash(&message);
-		let outcome =
-			XcmExecutor::<TestConfig>::execute_xcm(origin, message.into(), hash, 2_000_000_000_000);
+		let outcome = XcmExecutor::<TestConfig>::execute_xcm(
+			origin,
+			message.into(),
+			hash,
+			Weight::from_ref_time(2_000_000_000_000).set_proof_size(DEFAULT_PROOF_SIZE),
```
Comment (gavofyork) [inline xcm/xcm-builder/src/tests/bridging/mod.rs:140]:
```suggestion
			Weight::from_parts(2_000_000_000_000, 2_000_000_000_000),
```

File: xcm/src/v2/mod.rs (hunk 1 above)
Comment (KiChjang) [reply xcm/src/v2/mod.rs:900]: The `GetWeight` trait has changed, so this implementation would unfortunately have to change accordingly as well.

File: runtime/kusama/src/xcm_config.rs [hunk 10]
```diff
@@ -90,7 +91,7 @@ type LocalOriginConverter = (
 
 parameter_types! {
 	/// The amount of weight an XCM operation takes. This is a safe overestimate.
-	pub const BaseXcmWeight: u64 = 1_000_000_000;
+	pub const BaseXcmWeight: Weight = Weight::from_ref_time(1_000_000_000);
```
Comment (gavofyork) [inline runtime/kusama/src/xcm_config.rs:94]: should be `from_parts` and use a sensible upper limit for the amount of proof size an XCM instruction could reasonably take.

File: runtime/rococo/src/xcm_config.rs [hunk 11]
```diff
@@ -77,7 +78,7 @@ type LocalOriginConverter = (
 
 parameter_types! {
 	/// The amount of weight an XCM operation takes. This is a safe overestimate.
-	pub const BaseXcmWeight: u64 = 1_000_000_000;
+	pub const BaseXcmWeight: Weight = Weight::from_ref_time(1_000_000_000);
```
Comment (gavofyork) [inline runtime/rococo/src/xcm_config.rs:81]: I'd like to get rid of this completely, but if it stays, it should use `from_parts` with a conservative but sensible proof-size component.

File: xcm/pallet-xcm/src/mock.rs [hunk 12]
```diff
@@ -262,7 +263,7 @@ type LocalOriginConverter = (
 );
 
 parameter_types! {
-	pub const BaseXcmWeight: u64 = 1_000;
+	pub const BaseXcmWeight: Weight = Weight::from_ref_time(1_000);
```
Comment (gavofyork) [inline xcm/pallet-xcm/src/mock.rs:266]: ...as above

File: xcm/pallet-xcm/src/tests.rs [hunk 13]
```diff
@@ -65,8 +65,8 @@ fn report_outcome_notify_works() {
 				SetAppendix(Xcm(vec![ReportError(QueryResponseInfo {
 					destination: Parent.into(),
 					query_id: 0,
-					max_weight: 1_000_000
-				}),])),
+					max_weight: Weight::from_ref_time(1_000_000),
```
Comment (gavofyork) [inline xcm/pallet-xcm/src/tests.rs:68]: use `from_parts`

File: xcm/pallet-xcm/src/tests.rs [hunk 14]
```diff
@@ -82,13 +82,17 @@ fn report_outcome_notify_works() {
...
 		}]);
 		let hash = fake_message_hash(&message);
-		let r =
-			XcmExecutor::<XcmConfig>::execute_xcm(Parachain(PARA_ID), message, hash, 1_000_000_000);
-		assert_eq!(r, Outcome::Complete(1_000));
+		let r = XcmExecutor::<XcmConfig>::execute_xcm(
+			Parachain(PARA_ID),
+			message,
+			hash,
+			Weight::from_parts(1_000_000_000, 1_000_000_000),
```
Comment (gavofyork) [inline xcm/pallet-xcm/src/tests.rs:93]: 👍

File: xcm/pallet-xcm/src/tests.rs [hunk 15]
```diff
@@ -236,17 +244,17 @@ fn custom_querier_works() {
...
 		let hash = fake_message_hash(&message);
 		let r = XcmExecutor::<XcmConfig>::execute_xcm(
 			AccountId32 { network: None, id: ALICE.into() },
 			message,
 			hash,
-			1_000_000_000,
+			Weight::from_parts(1_000_000_000, 1_000_000_000),
 		);
-		assert_eq!(r, Outcome::Complete(1_000));
+		assert_eq!(r, Outcome::Complete(Weight::from_ref_time(1_000)));
```
Comment (gavofyork) [inline xcm/pallet-xcm/src/tests.rs:257]: `from_parts`

**Comment by gavofyork** [comment]
There should be no usages of `from_ref_time` in XCM code any more. Both dimensions are now of equal importance so use `from_parts` instead. If it's test code, you can make both dimensions the same, but don't forget to introduce tests which check for correct behaviour when they're different.

File: xcm/pallet-xcm/src/lib.rs [hunk 16]
```diff
@@ -699,7 +699,7 @@ pub mod pallet {
 							xcm: Xcm(vec![]),
 						},
 					]);
-					T::Weigher::weight(&mut message).map_or(Weight::MAX, |w| Weight::from_ref_time(100_000_000.saturating_add(w)))
```
Comment (KiChjang) [inline xcm/pallet-xcm/src/lib.rs:702]: @gavofyork The XCM pallet still contains calls to `from_ref_time` in places such as this. The intent here is to account for the weight used for executing the code in the extrinsic _without_ the weight from executing the XCM instructions. I suspect that this is not trivial to benchmark, hence the pseudo value of 100m being used here. As such, I don't believe that it makes sense to use `from_parts` here, but am unsure what the proper fix is, or whether I should just leave it alone.

Comment (gavofyork) [reply xcm/pallet-xcm/src/lib.rs:702]: This function needs weighing properly as the 100_000_000 is just a conservative guess.

**Comment by KiChjang** [comment]
/cmd queue -c bench-bot $ pallet westend pallet_xcm

**Comment by KiChjang** [comment]
/cmd queue -c bench-bot $ runtime westend pallet_xcm

**Comment by KiChjang** [comment]
/cmd queue -c bench-bot $ runtime westend-dev pallet_xcm

**Comment by KiChjang** [comment]
/cmd queue -c bench-bot $ runtime westend-dev pallet_xcm

File: xcm/pallet-xcm/src/lib.rs [hunk 17]
```diff
@@ -786,10 +849,12 @@ pub mod pallet {
...
 				hash,
-				max_weight, //.ref_time(),
-				max_weight, //.ref_time(),
+				max_weight,
+				max_weight,
 			);
-			let result = Ok(Some(outcome.weight_used().saturating_add(100_000_000)).into());
+			let result =
+				Ok(Some(outcome.weight_used().saturating_add(Weight::from_ref_time(100_000_000)))
+					.into());
```
Comment (gavofyork) [inline xcm/pallet-xcm/src/lib.rs:857]:
```suggestion
			let result = Ok(Some(outcome.weight_used().saturating_add(T::WeightInfo::execute())).into());
```

Comment (gavofyork) [reply xcm/pallet-xcm/src/lib.rs:857]: Note we should not be seeing *any* hard-coded weight values *at all* with this PR.

File: xcm/pallet-xcm/src/mock.rs [hunk 18]
```diff
@@ -77,7 +78,7 @@ pub mod pallet_test_notifier {
 
 	#[pallet::call]
 	impl<T: Config> Pallet<T> {
-		#[pallet::weight(1_000_000)]
+		#[pallet::weight(Weight::from_parts(1_000_000, 1_000_000))]
```
Comment (gavofyork) [inline xcm/pallet-xcm/src/mock.rs:81]: needs proper weighing.

File: xcm/pallet-xcm/src/mock.rs [hunk 19]
```diff
@@ -92,7 +93,7 @@ pub mod pallet_test_notifier {
 			Ok(())
 		}
 
-		#[pallet::weight(1_000_000)]
+		#[pallet::weight(Weight::from_parts(1_000_000, 1_000_000))]
```
Comment (gavofyork) [inline xcm/pallet-xcm/src/mock.rs:96]: ditto.

File: xcm/pallet-xcm/src/mock.rs [hunk 20]
```diff
@@ -113,7 +114,7 @@ pub mod pallet_test_notifier {
...
```
Comment (gavofyork) [inline xcm/pallet-xcm/src/mock.rs:117]: ditto.

**Review by gavofyork — APPROVED** [review]
Notwithstanding my comments, approved.

**Comment by KiChjang** [comment]
/cmd queue -c bench-bot $ runtime westend-dev pallet_xcm

**Comment by KiChjang** [comment]
/cmd queue -c bench-bot $ runtime rococo-dev pallet_xcm

**Comment by KiChjang** [comment]
/cmd queue -c bench-bot $ runtime kusama-dev pallet_xcm

**Comment by KiChjang** [comment]
/cmd queue -c bench-bot $ runtime polkadot-dev pallet_xcm

**Comment by KiChjang** [comment]
/cmd queue -c bench-bot $ runtime westend-dev pallet_xcm

**Comment by KiChjang** [comment]
/cmd queue -c bench-bot $ runtime westend-dev pallet_xcm

**Comment by KiChjang** [comment]
/cmd queue -c bench-bot $ runtime rococo-dev pallet_xcm

**Comment by KiChjang** [comment]
/cmd queue -c bench-bot $ runtime polkadot-dev pallet_xcm

**Comment by KiChjang** [comment]
/cmd queue -c bench-bot $ runtime westend-dev pallet_xcm

**Comment by KiChjang** [comment]
/cmd queue -c bench-bot $ runtime kusama-dev pallet_xcm

**Comment by KiChjang** [comment]
/cmd queue -c bench-bot $ runtime polkadot-dev pallet_xcm

**Comment by KiChjang** [comment]
/cmd queue -c bench-bot $ runtime polkadot-dev pallet_xcm

**Comment by KiChjang** [comment]
/cmd queue -c bench-bot $ runtime polkadot-dev pallet_xcm

**Comment by KiChjang** [comment]
/cmd queue -c bench-bot $ runtime polkadot-dev pallet_xcm

**Comment by KiChjang** [comment]
/cmd queue -c bench-bot $ runtime polkadot-dev pallet_xcm

**Comment by KiChjang** [comment]
/cmd queue -c bench-bot $ runtime polkadot-dev pallet_xcm

**Comment by KiChjang** [comment]
/cmd queue -c bench-bot $ runtime polkadot-dev pallet_xcm

**Comment by KiChjang** [comment]
/cmd queue -c bench-bot $ runtime polkadot-dev pallet_xcm

**Comment by KiChjang** [comment]
/cmd queue -c bench-bot $ runtime westend-dev pallet_xcm

**Pull request merged**